clients/apple/Shared/Stateful Logic/BillingService.swift

```diff
@@ -0,0 +1,197 @@
+        return Task.detached {
+            for await verification in Transaction.updates {
+                do {
+                    let transaction = try self.checkVerified(verification)
+                    switch self.core.getUsage() {
+                    case .success(let usages):
+                        if usages.dataCap.exact == 1000000 {
+                            if let receipt = self.getReceipt(), !self.makingPurchaseAttempt, transaction.id == transaction.originalID {
+                                guard let appAccountToken = transaction.appAccountToken else {
+                                    throw StoreError.failedVerification
+                                }
+                                
+                                let result = self.core.upgradeAccountAppStore(originalTransactionId: String(transaction.originalID), appAccountToken: appAccountToken.uuidString.lowercased(), encodedReceipt: receipt)
+                                
+                                switch result {
+                                case .success(_):
+                                    await transaction.finish()
+                                    self.purchaseResult = .success
+                                case .failure(let error):
+                                    DI.errors.handleError(error)
+                                }
+                            }
+                        } else {
+                            await transaction.finish()
+                        }
+                    case .failure(let error):
+                        DI.errors.handleError(error)
+                    }
+                } catch {
+                    print("Transaction failed verification")
+                }
+            }
+        }
+    }
+
+    
+    func purchasePremium() async throws -> PurchaseResult? {
+        let accountToken = UUID()
+        let purchaseOpt: Set<Product.PurchaseOption> = [Product.PurchaseOption.appAccountToken(accountToken)]
+        
+        do {
+            guard let monthlySubscription = maybeMonthlySubscription else {
+                throw StoreError.noProduct
+            }
+            let result = try await monthlySubscription.purchase(options: purchaseOpt)
+            SKReceiptRefreshRequest().start()
+            
+            switch result {
+            case .success(let verification):
+                while true {
+                    if let receipt = getReceipt() {
+                        makingPurchaseAttempt = true
+                        let transaction = try checkVerified(verification)
+                        
+                        let result = self.core.upgradeAccountAppStore(originalTransactionId: String(transaction.originalID), appAccountToken: accountToken.uuidString.lowercased(), encodedReceipt: receipt)
+                        
+                        switch result {
+                        case .success(_):
+                            await transaction.finish()
```

Comment:
pretty nested at this point, I think there's probably a way to write this all to be more readable